### PR TITLE
ssh-agent Signer

### PR DESCRIPTION
**Pull request by kim**
Worked example of how to use an `ssh-agent(2)` as a `Signer` impl.

There are some limitations of the `thrussh-keys` crate, however:

* The signature of `AgentClient::sign_request_signature` is probably
  wrong, which leads to the requirement of setting the `SSH_AUTH_SOCK`
  environment variable
* The constraints types are either wrong, or the standard distro agent
  doesn't understand them (or not all of them, eg. KeyLifetime)
* It is unfortunate that `sodiumoxide` and `tokio` get pulled in.
  Perhaps Pierre can be convinced of the idea to abstract over the former,
  at least.

Signed-off-by: Kim Altintop <kim@monadic.xyz>

### CODE REVIEW

**Review by FintanH — APPROVED** [review]
:bird:

**Review by xla — COMMENTED** [review]
>     * The signature of `AgentClient::sign_request_signature` is probably
>       wrong, which leads to the requirement of setting the `SSH_AUTH_SOCK`
>       environment variable

Isn't `SSH_AUTH_SOCK` standard practice for a setup like this? It sounds like integrating on it is sub-optimal, or am I misreading your point?

>     * It is unfortunate that `sodiumoxide` and `tokio` get pulled in.
>       Perhaps Pierre can be convinced of the idea to abstract over the former,
>       at least.

Aren't those in `std` yet?

trolololo

**Comment by kim** [comment]
> Isn't `SSH_AUTH_SOCK` standard practice for a setup like this? It sounds like integrating on it is sub-optimal, or am I misreading your point?

Sure yes, but you’d want to not have this baked into a library, because, environment variables are the worst API you can have. 

The problem here is that the signature consumes the socket when an error occurs, even though the socket might still be fine. We would either need to change the trait signature to also consume `self`, so as to force users to throw away the signer when an error occurs, or we just try to reconnect. Unfortunately, we can’t write `impl Fn() -> Result<impl AsyncRead + AsyncWrite, Error>`.

File: examples/ssh-agent.rs [hunk 1]
```diff
@@ -0,0 +1,57 @@
+// This file is part of radicle-link
+// <https://github.com/radicle-dev/radicle-link>
```
[the remark this answers is not in the document]
Comment (kim) [reply examples/ssh-agent.rs:2]: Looks like _all_ license headers are wrong. Should probably be fixed in its own patch.

File: Cargo.toml [hunk 2]
```diff
@@ -17,14 +20,15 @@ rpassword = "4.0"
 secstr = "0.3"
 serde = { version = "1.0", features = ["derive"] }
 serde_cbor = "0.10"
+thrussh-keys = { version = "0.20", optional = true }
+tokio = { version = "1.0", features = ["net"], optional = true }
```
[the remark this answers is not in the document]
Comment (kim) [reply Cargo.toml:24]: It is possible that I’m misunderstanding something about cargo’s dependency solver, the docs are wrong, or the cargo defaults are just the worst possible choice.

Afaiu, the above is equivalent to “^1.0”, which supposedly means that cargo can use a newer version if it starts with “1.0”, eg 1.0.3, 1.0.10, etc

This is mucho semantico, because the dependency is supposed to only increment the zero if it would somehow break consumers which expect the previous API. Right?


Now, what I see a lot is that crates actually depend on a three-digit version, eg `quinn` depends on `tokio ^1.0.1`. What I would expect to happen if I set this to `1.4` here, and then use this crate from librad is that I’ll end up with two versions of tokio (which is a situation also referred to as “async hell”).

Is there some semantic wishywashy I’m missing? Or is it just really bad practice to have such strict version bounds?

[the remark this answers is not in the document]
Comment (kim) [reply Cargo.toml:24]: Yeah, so what I was missing is the difference between versions that start with zero, and ones that don’t (zeroes are special).

```
^1.2.3  :=  >=1.2.3, <2.0.0
^1.2    :=  >=1.2.0, <2.0.0
^1      :=  >=1.0.0, <2.0.0
^0.2.3  :=  >=0.2.3, <0.3.0
^0.2    :=  >=0.2.0, <0.3.0
^0.0.3  :=  >=0.0.3, <0.0.4
^0.0    :=  >=0.0.0, <0.1.0
^0      :=  >=0.0.0, <1.0.0
```

**tl;dr** Since tokio is > 1.0, it doesn’t really matter. But using `1.4` here will auto-upgrade all consumers.

[the remark this answers is not in the document]
Comment (kim) [reply Cargo.toml:24]: I suppose a superior software versioning system would be to use the hash of the AST on module level, and then specify in your dependencies file how many **leading zeroes** is shall have.

**Comment by kim** [comment]
Wondering if @P-E-Meunier has a way to accept patches while the Nest is being rebuilt from the ashes. I suppose I could just serve the `thrussh` repo over HTTP somewhere, but the snapshot I have is from Feb 2020.

**Comment by P-E-Meunier** [comment]
The Nest will come back to life today, I could finally access the backups yesterday. The repositories are already converted to the new repository format, the only thing that remains to be done is the high availability layer for the repositories.

Other than that, I've briefly looked at this issue. Remember that the main concern with Thrussh is that everybody has strong (and different) expectations about how SSH should behave. For example, many people use outdated crypto (weak RSA protected by old key generation schemes, or PGP), yet expect strong security from the library, no external dependencies, etc.

**Comment by kim** [comment]
On Wed Mar 24, 2021 at 8:23 AM CET, Pierre-Etienne Meunier wrote:
> The Nest will come back to life today

That's great to hear!

> For example, many people use outdated crypto (weak RSA protected by old key
> generation schemes, or PGP), yet expect strong security from the library, no
> external dependencies, etc.

Yes, that's good points.

I think the thing that should be easy to fix (or clarify why it is the way it
is) is the signature of `AgentClient::sign_request_signature` -- since
`sign_request` and `sign_request_base64` return `Self` even in case of error,
this must be an oversight.

About sodium and tokio I would think that there's a way to give users a choice
via features.

**Review by FintanH — APPROVED** [review]
Looks good to me :smirk:
